### DISCUSSION TITLE
Problem with predictions using Gaussian Noise defence

### DISCUSSION THREAD

**Discussion by GianlucaMega** [Trusted-AI/adversarial-robustness-toolbox#1674] [Q&A] [answered]
Hello,
I'm trying to use Gaussian noise as postprocessing defence but I'm getting an error when calling the predict method of the classifier. I'm using the code from the keras example of this repo which I'm reporting here:

```
import tensorflow as tf
tf.compat.v1.disable_eager_execution()
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Conv2D, MaxPooling2D
from tensorflow.keras.losses import categorical_crossentropy
from tensorflow.keras.optimizers import Adam
import numpy as np
from art.attacks.evasion import FastGradientMethod
from art.estimators.classification import KerasClassifier
from art.utils import load_mnist, compute_accuracy
from art.defences.postprocessor import GaussianNoise

(x_train, y_train), (x_test, y_test), min_pixel_value, max_pixel_value = load_mnist()

model = Sequential()
model.add(Conv2D(filters=4, kernel_size=(5, 5), strides=1, activation="relu", input_shape=(28, 28, 1)))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Conv2D(filters=10, kernel_size=(5, 5), strides=1, activation="relu", input_shape=(23, 23, 4)))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(100, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(loss=categorical_crossentropy, optimizer=Adam(learning_rate=0.01), metrics=["accuracy"])

gauss = GaussianNoise()
classifier = KerasClassifier(model=model, clip_values=(min_pixel_value, max_pixel_value), use_logits=False, postprocessing_defences=gauss)

classifier.fit(x_train, y_train, batch_size=64, nb_epochs=3)
preds = classifier.predict(x_test)
acc = compute_accuracy(preds, y_test)
```

and this is the error that is prompted:
```
---------------------------------------------------------------------------
ValueError  Traceback (most recent call last) [<ipython-input-5-eec5e88d0d3f>](https://localhost:8080/#) in <module>()
     37 
     38 classifier.fit(x_train, y_train, batch_size=64, nb_epochs=3)
---> 39 preds = classifier.predict(x_test)
     40 acc = compute_accuracy(preds, y_test)

3 frames
[/usr/local/lib/python3.7/dist-packages/art/defences/postprocessor/gaussian_noise.py](https://localhost:8080/#) in __call__(self, preds)
     71                 post_preds[post_preds < 0.0] = 0.0
     72                 sums = np.sum(post_preds, axis=1)
---> 73                 post_preds /= sums
     74         else:
     75             post_preds[post_preds < 0.0] = 0.0

ValueError: operands could not be broadcast together with shapes (10000,10) (10000,) (10000,10)
```

Is this a bug or am I doing something wrong?

Thank you for the support

**Comment by beat-buesser** [comment, marked as answer]
Hi @GianlucaMega Thank you for raising this question! I think this is indeed a bug, the division in line 73 works only for a single sample. To be able to handle mini-batches line 73 should be `sums = np.sum(post_preds, axis=1, keepdims=True)`.

I'll transform this discussion into a bug report issue.

Would you be interested to open a pull request with the solution to branch `dev_1.10.2`?

**Reply by GianlucaMega** [reply]
Hi @beat-buesser, sorry for not answering you. I tried your solution locally and it seems to work but I don't have any experience in contributing or opening pull requests so I don't know how to do it.

**Reply by beat-buesser** [reply]
Hi @GianlucaMega I think this could be a great opportunity to get started! If you please follow these steps and let me know where you have questions:

1. Fork this repo by clicking on `Fork` in the top-right corner of this repo
2. Clone the newly forked repo
3. Switch the cloned repo locally to branch `dev_1.10.2`
4. Make the changes discussed above in the cloned repo
5. Create a commit of the changed file using `git add <filename> and `git commit -s`
6. Push your commit to your forked repo with `git push origin dev_1.10.2`
7. Open your forked repo in a web-browser and follow these steps to open a pull request. You will have to change
- the compare branch to your `dev_1.10.2`
- the base branch to `dev_1.10.2`
- the base repository to `Trusted-AI/adversarial-robustness-toolbox`
- https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request

**Reply by GianlucaMega** [reply]
Alright then, I'll do it today. Thanks again for the support!

**Reply by beat-buesser** [reply]
That's amazing, thank you very much!
